pkg/generate/app/cmd/newapp.go

```diff
@@ -314,10 +348,13 @@ func (c *AppConfig) Run(out io.Writer) (*AppResult, error) {
 	glog.V(4).Infof("Code %v", repositories)
 	glog.V(4).Infof("Images %v", components)
 
-	if err := c.detectSource(repositories); err != nil {
+	srcComponents, err := c.detectSource(repositories)
+	if err != nil {
```

Comment:
Regarding the second part of your comment... the component is only added if the source hasn't been matched up to one of the components you specified. (https://github.com/openshift/origin/pull/1150/files#diff-67f42376bdba60c035c6268d160059baR199). 

And yes, we want to do source detection first. The trick is that we want to conditionally include the component only if we can't match the source to one of the components you specified. I wasn't thinking of solving that problem yet with this PR. :-)